### DISCUSSION TITLE
COMSOL Version Number

### DISCUSSION THREAD

**Discussion by zkb17** [wmglab-duke/ascent#25] [Q&A] [answered]
Currently, my university can only provide free access to COMSOL v5.3, not v5.5, as is prescribed in the latest installation instructions. Does anyone know if this pipeline still works with COMSOL v5.3 or with other previous version of COMSOL?

**Comment by ericmusselman** [comment, marked as answer]
Hi @zkb17

We just tested our beta task with version COMSOL v5.3 to check. Unfortunately, we learned that we use some new functionality in COMSOL v5.4+ and ASCENT will not work with COMSOL v5.3. Would it be possible for you to have access to v5.4?

Specifically, we are getting errors that v5.3 does not have the "ModelParamGroup" class that was introduced in v5.4.

![image](https://user-images.githubusercontent.com/18255798/165323060-830c7d3c-45f7-4077-bb17-073e87dd99b3.png)

**Reply by zkb17** [reply]
@ericmusselman Thanks for getting back to me on this one. I have been so confused the past week about what I was doing wrong. I was getting an error with the model parameter as well, so now it makes sense at least. Sadly, I don't have an immediate way to obtain v5.4+ from my university, but I will see what I can do. ASCENT is the most well suited simulation for my project that I've found out there, so I am going to try and procure a newer version of COMSOL.

**Reply by zkb17** [reply]
@ericmusselman I ran it again today on v5.3 to see if the error was the same (just to verify), but it seems that the error is slightly different. It occurs at the same point, after java initiates an instance of COMSOL, and it has something to do with the ModelWrapper as well, but slightly different. I was wondering if you would take a look at this error and see if you think it is due to the version error. I just want to ensure I am not doing something ELSE wrong as well. It is also worth nothing that the program continued to run in mine and it exported files that could be sent to NEURON. However, those files are empty. 
![Error Screen for GitHub - ModelWrapper Class](https://user-images.githubusercontent.com/104382565/165368182-d4c0741c-a5f4-4ff7-b8a0-9c659c6ec940.JPG)

**Comment by ericmusselman** [comment]
Did you change your JDK installation? This looks like an error in Runner.py's handoff() method, where a call is made with javac to compile the Java files. To get the error I posted earlier, this would have had to be correct at the time of running the pipeline.

On my Windows machine, which it appears you are using as well, the env.json file (in config/system/) looks like the following:

{
  "ASCENT_COMSOL_PATH": "C:\\Program Files\\COMSOL\\COMSOL54\\Multiphysics",
  "ASCENT_JDK_PATH": "C:\\Program Files\\Java\\jdk1.8.0_221\\bin",
  "ASCENT_PROJECT_PATH": "D:\\Documents\\ascent",
  "ASCENT_NSIM_EXPORT_PATH": "D:\\Documents\\ascent\\submit"
}

Please check that your "ASCENT_JDK_PATH" is similar.

**Reply by zkb17** [reply]
I believe I did all the env.json file correctly. Please check below. 

The only other thing I have done with the model component of the program is copy the model.json file from templates to the correct file path as indicated in documentation. The documentation then goes on to say that users should edit the model.json or make their own preset name.json model file to design their own cuff. I have left the model.json file from the template as in for the first run and I used the beta task model.json file for the beta task run. Neither the "first run" nor the beta task pipelines work for me. Same error in both. 

{
  "ASCENT_COMSOL_PATH": "C:\\Program Files\\COMSOL\\COMSOL53a\\Multiphysics",
  "ASCENT_JDK_PATH": "C:\\Program Files\\Java\\jdk1.8.0_221\\bin",
  "ASCENT_PROJECT_PATH": "C:\\Users\\zackb\\OneDrive\\Documents\\ICL\\MRes Bioengineering\\ASCENT\\ascent-master",
  "ASCENT_NSIM_EXPORT_PATH": "C:\\Users\\zackb\\OneDrive\\Documents\\ICL\\MRes Bioengineering\\ASCENT"
}

**Comment by zkb17** [comment]
I have managed to get the upgrade to v5.5 from COMSOL. I reran the ASCENT pipeline, but I am still getting the same error that I showed earlier. My env.json file is correct, as far as I can tell. It looks the same as yours with my file directories. Do you have any other recommendations to try? I will go slowly back through the installation process and file allocation process described in the "first run" section of the documentation, but I cannot tell what I am doing wrong to get this error.

**Comment by ericmusselman** [comment]
Hi @zkb17 

Which error, this one?

error: Invalid filename: model\\*.java

To help us look into the issue, which version of Windows are you running?

**Comment by zkb17** [comment]
Yes, same error as the previous one that I sent an image of.

Windows 10 Pro v10.0.19044

**Comment by zkb17** [comment]
Hello @ericmusselman 

I have done some digging into the runner.py handoff() method. So the error arises at the highlighted section in the runner.py script. I have included pictures of the error I am getting as well. I just added in print() commands to diagnose where exactly it was occuring. But in the print() commands, you can also see what I using for my jdk_path, comsol_path, and project_path, which is helpful. 

Regarding the script in runner.py handoff() method, the issue lies in that first os.system() call. The argument of that method appears to be setting the classpath (-cp) for javac to jre.jar file in the ascent-master/bin directory and also to the comsol plugins, which is fine, I think. and then it tries to run the *.java files when it executes model//*.java, but it can't find those files it says... well those files aren't in either of the classpaths that we just told javac to look to find the source files, right? The path for the *.java files is ../ascent_master/src/model/*.java. Now, I am running everything from the ../ascent_master directory, but don't we need to tell javac to also look in the /src/model directory for the *.java files? Because right now we are setting the classpaths (-cp) to paths that don't contain those *.java files. Perhaps I am misunderstanding something, but what are your thoughts?

![image](https://user-images.githubusercontent.com/104382565/166978135-08c2504d-a48d-449f-87b3-31ef01f38c69.png)
![image](https://user-images.githubusercontent.com/104382565/166984347-00877f37-d546-4d11-8f5a-8703505d9f9c.png)

**Comment by ericmusselman** [comment]
Hi @zkb17 

Can you please add the following lines just before the "javac" line, and post the traceback here?

```
print(os.getcwd())
print(os.listdir('model'))
```

Thanks!

PS: your error changed from "invalid filename" to "file not found" -- so something changed on your end.

**Reply by zkb17** [reply]
@ericmusselman Here is the traceback like you asked for after adding those lines of code... it does appear to going into src and the .java files are there... but is the -cp command within the os.system() method not redirecting the classpath to something besides cwd?

![image](https://user-images.githubusercontent.com/104382565/167021308-16cd5e18-f0f3-4eb8-ab79-22ebfa4adf36.png)

**Reply by thedannymarsh** [reply]
Hi,

We're still not sure what is going on. The classpath is different from source files, which is the target for the compiler. (https://docs.oracle.com/javase/7/docs/technotes/tools/windows/javac.html)
As a stopgap, could you replace
`model\\*.java`
with 
`IdentifierManager.java JSONio.java ModelSearcher.java ModelWrapper.java Part.java`
This should allow you to compile the code properly and run the pipeline.

Thanks!
-Daniel

**Comment by zkb17** [comment]
@dmarshall-bing @ericmusselman First of all, thank you both for getting back to me on this issue and trying to help. You wrote a whole documentation on the pipeline to avoid this, but alas, here I am. 

I tried the stopgap measure that you suggested, and it still doesn't work. It gives the same error, but instead "javac: file not found: model\\*.java" it is now "javac: file not found: IdentifierManager.java". So it can't file that first file listed so it throws an error.

I think there is still something amiss with the -cp option. I understand that classpath is different from where the source files are, but in the documentation that you sent me, it states that if the -sourcepath variable is not set, then the compiler with search for the source files in the -classpath variable if it is given. So in this case, I think javac is searching for source files in the two classpaths that we gave it, the jre.jar file and the comsol\plugins folder. Which would mean that there is no .java file to be found in either of those cps, so the error makes sense. ? What do you think about that? 

Also, please don't feel like you need to continue helping me troubleshoot this. I am going to try and keep at it to see what is up. I think your pipeline is quite powerful and very useful for my project, so I am going to put some time into it.

**Reply by thedannymarsh** [reply]
We are happy to help! The classpath is likely not the issue here, as the code runs OK on our machines (where we are also running Windows 10). I would suggest three steps to try to troubleshoot at this point:
1. Instead of just manually specifying just the filenames, manually specify the absolute path (e.g. replace ModelWrapper.java with C:\\Users\\<rest of path>\\ascent-master\\src\\model\\ModeWrapper.java).
2. Try passing the command directly to the system shell (i.e. open cmd and input the command. you can get the command text by replacing the os.system call with a print call and copying the printout).
3. If neither of these work, the issue is likely with your java installation. I would try reinstalling java and seeing if that fixes the issue. Let us know how it goes!
Thanks!
-Daniel
